Tech-Fellows-SnoopJake/EmailServer
# HG0 Improvements  --> HG5

**WHY THIS PR?**
This PR enhances user experience with registration, logout, authentication, and also strengthens security by refining secret key management and code practices based on SonarCloud insights.

**WHAT CHANGED?**

_Backend_
Secret Key Management: Transitioned from reading the backend secrets keys from a text file on the EC2 instance to securely storing it within the GitHub organization's secrets. This allows the workflow to access the secret keys during the deployment process, enhancing the application's security.
settings.py Update: removed the hardcoded SECRET_KEY and updated the code to load the SECRET_KEY from environment variables
views.py Update: 
Class name standardization to align with PEP 8's CapWords convention, improving code readability and consistency.
Enable user registration and user authentication: refactor logic to create application users and safely secure the access to the API, using the django auth users implementation.
Use JWT tokens: authenticate users that access from the frontend client application.

_Frontend_
API URL Secret Key: Updated the API URL to utilize the secret key method, ensuring secure communication with the backend.
Fix code to use JWT tokens authentication
Update interface for registration and login features.
Front-end route protection, to restrict users from navigating without being logged in.
Code Refactoring:
Addressed issues highlighted by SonarCloud, specifically fixing multiple imports in Login.tsx and App.tsx.
Refactored the LOGIN function to reduce cognitive complexity, making the code more readable and maintainable.

_Deployment_
Database Migration: Redeployed the application with a MySQL database, including necessary code updates to accommodate this change, enhancing the application's scalability and performance.
Maintained the workflow process through the ci/cd pipeline, but adapted the environment variables and processes to deploy to our resources.

**TODO**
Future Tasks: Continue monitoring the security and performance of secret key management and database integration to ensure they meet our evolving requirements.
Codebase Maintenance: Regularly refactor and review the codebase to maintain high standards of code quality and consistency.

**QA / VALIDATION STEPS**
Testing Secret Key Management: Verified that the new secret key management process works correctly by deploying the application and ensuring that it starts without issues, confirming that the secret key is appropriately loaded from GitHub secrets.
Frontend Validation: Tested the frontend changes by navigating through the application, ensuring that API calls are successful and that the UI behaves as expected after the refactor.
Database Migration Test: Confirmed the successful migration to a MySQL database by performing CRUD operations on the database through the application and verifying data integrity and performance.
Use sonar cloud feedback to analyze the code structure and follow more good practices recommendations.

**Note:** The README has been updated to reflect changes in secret key management and deployment instructions to guide new and existing contributors on how to set up their development and production environments securely.
